src/org/sosy_lab/cpachecker/cfa/mutation/strategy/GenericCFAMutationStrategy.java

```diff
@@ -0,0 +1,368 @@
+  }
+
+  protected boolean alreadyTried(ObjectKey pObject) {
+    return previousPasses.contains(pObject) || previousMutations.contains(pObject);
+  }
+
+  protected abstract RollbackInfo getRollbackInfo(ParseResult pParseResult, ObjectKey pObject);
+
+  protected abstract void removeObject(ParseResult pParseResult, ObjectKey pObject);
+
+  protected abstract void returnObject(ParseResult pParseResult, RollbackInfo pRollbackInfo);
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public boolean mutate(ParseResult pParseResult) {
+    currentMutation.clear();
+
+    switch (state) {
+      case NewLevel:
+        if (!setLevel(pParseResult)) {
+          return false;
+        }
+        if (rate <= 2) {
+          state = State.RemoveDelta;
+          return mutate(pParseResult);
+        }
+        state = State.RemoveComplement;
+        // $FALL-THROUGH$
+
+      case RemoveComplement:
+        if (batchNum++ < batchCount) {
+          currentMutation.addAll(getObjects(pParseResult, levelSize - batchSize));
+          logger.logf(
+              Level.INFO,
+              "removing complement %d/%d: %d %s",
+              batchNum,
+              batchCount,
+              currentMutation.size(),
+              objectsDescription);
+          if (!currentMutation.isEmpty()) {
+            state = State.RemoveComplementNoRollback;
+            break;
+          }
+        }
+        batchNum = 0;
+        state = State.RemoveDelta;
+        previousMutations.clear();
+        logger.logf(Level.INFO, "switching to deltas");
+        // $FALL-THROUGH$
+
+      case RemoveDelta:
+        if (batchNum++ < batchCount) {
+          currentMutation.addAll(getObjects(pParseResult, batchSize));
+          logger.logf(
+              Level.INFO,
+              "removing delta %d: %d %s",
+              batchNum,
+              currentMutation.size(),
+              objectsDescription);
+          if (!currentMutation.isEmpty()) {
+            previousMutations.addAll(currentMutation);
+            state = State.RemoveDeltaNoRollback;
+            break;
+          }
+        }
+        state = State.NewLevel;
+        return mutate(pParseResult);
+
+      case RemoveComplementNoRollback:
+        logger.logf(Level.INFO, "complement %d/%d was removed successfully", batchNum, batchCount);
+        rate = 2;
+        state = State.NewLevel;
+        return mutate(pParseResult);
+
+      case RemoveDeltaNoRollback:
+        logger.logf(Level.INFO, "delta %d/%d was removed successfully", batchNum, batchCount);
+        state = --rate < 2 ? State.NewLevel : State.RemoveDelta;
+        return mutate(pParseResult);
+    }
+
+    stats.rounds.inc();
+    rollbackInfos.clear();
+    for (ObjectKey object : currentMutation) {
+      final RollbackInfo ri = getRollbackInfo(pParseResult, object);
+      rollbackInfos.push(ri);
+      //      logger.logf(Level.INFO, "removing %s with ri %s", object, ri);
+      removeObject(pParseResult, object);
```

Comment:
Maybe, removeObject should return RollabckInfo? Then a separate method getRollbackInfo is not necessary.

Reply:
It's a good idea.